SBI: Attestation extension

sameo opened the pull request:
Signed-off-by: Samuel Ortiz <sameo@rivosinc.com>

sameo commented:
cc @rsahita

sameo commented:
* Call it measurement and attestation
* csr vs RISCV CSR
* Non normative paragraph for flow description
* Get evidence error: try again/busy
* Enumeration function (platform capabilities, mandate+supported)
  * Hash types
  * Evidence extensions format
  * TCB SVN
  * Measurment registers indexes
  * (TCB firmware updateable?) (Update Notifications?)

sameo commented:
cc @atulkharerivos

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+* A X.509 extension formatted according to `evidence_format`, containing the
+attestation evidence that is associated with the CSR `Subject` key. In other
+words, the generated certificate `Subject` and `SubjectPublicKey` identify the
+entity to which this extension defined evidence applies to.
```
Avoid the dangling preposition by using something like "...identify the entity associated with this extension's defined evidence". Also, might be possible to drop "In other words" with  ", i.e.,"

sameo replied:
Addressed. I used `In other words` as it was the beginning of the sentence.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+                                           unsigned long evidence_data_addr,
+                                           unsigned long evidence_format,
+                                           unsigned long cert_addr,
+                                           unsigned long max_cert_size)
```
Nit: Probably needs a ; at the end to be valid C-code.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+----
+struct sbiret sbi_measurement_extend(unsigned long measurement_addr,
+                                     unsigned long measurement_index,
+                                     unsigned long hash_algorithm)
```
Nit: Probably needs a ; at the end to be valid C-code; also applies to other places below.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+| TCB_SVN               | 0              | Integer        | 4                      | TCB Security Version Number
+| HASH_ALGORITHMS       | 4              | Bitmap         | 1                     a| Supported measurement hash algorithms bitmap:
+
+                                                                                     * Bit 0: SHA-384
```
Out of curiosity, is SHA-256 no longer being used for this purpose? The reason for asking is that TPM2 is still SHA256.

sameo replied:
SHA 256 is not secure against length extension attacks, and none of the existing TEE implementations (TDX, SEV, SE, etc) use it. They use SHA-384.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+  A <<measurement-extend-function>> call must pass one of the supported
+  algorithms.
+* `EVIDENCE_FORMATS`: A bitmap of all supported attestation evidence X.509
+extensions format. A <<get-evidence-function>> call must pass one of the
```
s/format/formats

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+* `EVIDENCE_FORMATS`: A bitmap of all supported attestation evidence X.509
+extensions format. A <<get-evidence-function>> call must pass one of the
+supported formats.
+* `NUM_SMSMT_REGS`: The number of static, non extensible, measurement registers.
```
s/non extensible/non-extensible

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+<<table_attestation_and_measurement_capabilities_format>>.
+
+It is the caller's responsibility to allocate enough memory at `caps_addr` to
+accomodate for the largest possible capabilities structure, which is 136 bytes.
```
"accommodate the largest possible capabilities structure (136 bytes)."

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+Get the SBI implementation attestation and measurement capabilities.
+
+The reported capabilities allow attestation and measurement implementations to
+build platform agnostic frameworks, by dynamically adapting the SBI calls to the
```
Consider terminating the sentence at "....platform agnostic frameworks.". The ", by dynamically..." seems a little redundant, and slightly confusing".

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+Extend the platform TCB measurement with one additional measurement.
+
+This function typically allows the caller to add a software component to the
```
Nit: "typically" seems redundant.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+This function typically allows the caller to add a software component to the
+list of the TCB layers measurements at any point of the platform lifecycle.
+By supporting this function, the SBI implementation implicitly becomes a
```
Consider rephrasing as "This is the equivalent of DRTM support in the SBI implementation." or something similar.

sameo replied:
I tried with `From an SBI implementation perspective, this function is equivalent to supporting the TCG DRTM capability.` Not sure it's better...

atulkharerivos replied:
How about flipping the perspective :), i.e.: "This function provides a DRTM-like capability by allowing the caller to add a software component to the list of the TCB layers measurements at any point of the platform lifecycle."

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+By supporting this function, the SBI implementation implicitly becomes a
+https://trustedcomputinggroup.org/wp-content/uploads/TCG_D-RTM_Architecture_v1-0_Published_06172013.pdf[Dynamic Root of Trust for Measurement (DRTM)].
+
+Measurements added through this SBI function should be managed separately from
```
Clarify actor through use of active voice: "The SBI implementation should manage measurement added through this SBI function separately. It can choose to..."

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+structure (`ATST_CAPS`). See
+<<table_attestation_and_measurement_capabilities_format>> for more details.
+
+The possible error codes returned in `sbiret.error` are shown in the
```
Nit: Consider active voice..."The <<table_measurement_extend_errors>> table below shows the possible error codes returned in `sbiret.error`"

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+* A https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.1[AuthorityKeyIdentifier]
+X.509 extension.
+
+The possible error codes returned in `sbiret.error` are shown in the
```
See note about active voice below.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+extensions format. A <<get-evidence-function>> call must pass one of the
+supported formats.
+* `NUM_SMSMT_REGS`: The number of static, non extensible, measurement registers.
+These are typically used by the platform to measure all boot components.
```
Prefer to be prescriptive by dropping "typically".

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+through a <<measurement-extend-function>> call. This is set to 0 if the platform
+does not allow for extending the TCB  measurement.
+* `MSMT_REGS_INFO`: An array of measurement register descriptors, as defined by
+<<table_measurement_register_format>>. The size of this array is dynamic as it
```
Nit: Replace "as it", with "..., and depends on  on the number of static and runtime registers.".

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,250 @@ The possible error codes returned in `sbiret.error` are shown in the
+or the `DiceMultiTcbInfo` format for reporting attestation evidences.
+* `TCG_PCR_INDEX`: The https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf[`TCG PCR`]
+index the measurement registers map to.
+* `RUNTIME_MSMT`: Define if the register is part of the static or runtime
```
/s/Define/Defines

sameo commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+|===
+| Field Name            | Offset (bytes) | Type           | Size (bytes)           | Description
+| TCB_SVN               | 0              | Integer        | 4                      | TCB Security Version Number
+| HASH_ALGORITHMS       | 4              | Bitmap         | 1                     a| Supported measurement hash algorithms bitmap:
```
This should probably be one single algorithm.

sameo commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+----
+struct sbiret sbi_measurement_extend(unsigned long measurement_addr,
+                                     unsigned long measurement_index,
+                                     unsigned long hash_algorithm);
```
If the platform only exposes one algorithm, this argument can be removed.

rsahita commented:
> * Call it measurement and attestation
> * csr vs RISCV CSR
> * Non normative paragraph for flow description
> * Get evidence error: try again/busy
> * Enumeration function (platform capabilities, mandate+supported)
>   
>   * Hash types
>   * Evidence extensions format
>   * TCB SVN
>   * Measurment registers indexes
>   * (TCB firmware updateable?) (Update Notifications?)

+
* specify # of measurement indices are configurable (and static or dynamic is upto implementation)
* add a tbd for a future sealing CDI
* add some prose on the expected usage - measure n+1, create and prov. key/cert for n+1, lock key n, transfer control to n+1
* any limits on validity of attestation CDI may be enforced by layer n for n+1

rsahita commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+be used by attestation verifiers, either locally or remotely.
+
+These functions provide the capabilities and features required to implement
+confidential computing, and also  measured and verified boot.
```
These functions provide the capabilities and features required to implement attestation for usages such as, confidential computing, measured boot and verified boot.

rsahita commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+----
+struct sbiret sbi_attestation_get_evidence(unsigned long cert_request_addr,
+                                           unsigned long cert_request_size
+                                           unsigned long report_data_addr,
```
```suggestion
                                           unsigned long request_data_addr,
```

rsahita commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+address, which length is `cert_request_size`.
+
+Additionally, the caller provides a 64 bytes data blob at the
+`report_data_addr` physical address. This blob will be included in the
```
```suggestion
`request_data_addr` physical address. This blob will be included in the
```

sameo replied:
Yes, thanks for the reminder.

rsahita commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+implementing TCB attestation flows and for extending TCB measurements. The SBI
+implementation composes those measurements into an attestation evidence that can
+be used by attestation verifiers, either locally or remotely.
+
```
```suggestion
                                  Layer n+1
                                 +--------------------------------------------+
                                 | Extend Measurements for layer n+2          |
                                 | Generate evidence cert (CDI) for layer n+2 |
                                 | Access-control current layer keys          |
                                 | Transfer control to layer n+2              |
                                 +-------^------------------------------------+
                                         |
  Layer n                                |
+--------------------------------------------+
| Extend Measurements for layer n+1          |
| Generate evidence cert (CDI) for layer n+1 |
| Access-control current layer keys          |
| Transfer control to layer n+1              |
+--------------------------------------------+
```

sameo replied:
Added a couple of diagrams in the intro section. Please let me know how it looks.

rsahita commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1675,259 @@ The possible error codes returned in `sbiret.error` are shown in the
+| SBI_ERR_FAILED      | The evidence could not be generated.
+| SBI_INVALID_ADDRESS | The CSR or the certificate adress is invalid.
+| SBI_INVALID_PARAM   | The CSR at `csr_addr` is invalid.
+|===
```
| SBI_ERR_BUSY   | The request failed due to resource busy error and may be retried.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+Both the constitution and definition of the TCB and its measurement are
+implementation specific.
+
+The caller provides a https://datatracker.ietf.org/doc/html/rfc5280[X.509]
```
Are there any other certificate format allowed ? If not, please use normative text.

sameo replied:
That's the only format we want to support, I made it normative now.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+implementation specific.
+
+The caller provides a https://datatracker.ietf.org/doc/html/rfc5280[X.509]
+formatted Certificate Signing Request (CSR) at the `cert_request_addr` physical
```
Avoid using CSR as an acronym as CSR also stands for control status register and more commonly understood acronym. Using CSR for something else may create confusion.

sameo replied:
Fixed, thanks.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+                        `sbiret.value`.
+| SBI_ERR_FAILED      | The evidence could not be generated.
+| SBI_INVALID_ADDRESS | The CSR or the certificate adress is invalid.
+| SBI_INVALID_PARAM   | The CSR at `csr_addr` is invalid.
```
Other possible case where INVALID_PARAM can be returned ?
1. Invalid evidence format

Just FYI: you can always new error codes if required to provide more information why evidence could not be generated.

sameo replied:
The other possible error I could think about is that the generated evidence is larger than `cert_size`. So I added a `SBI_ERR_OVERFLOW` error code to express that.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+capability by allowing the caller to add a software component to the list of the
+TCB layers measurements at any point of the platform lifecycle.
+
+The SBI implementation should manage measurement added through this SBI function
```
must ? Any non-normative text should be avoided if the intent is not

sameo replied:
It's normative now.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+[source, C]
+----
+struct sbiret sbi_measurement_extend(unsigned long measurement_addr,
```
/s/sbi_measurement_extend/sbi_attestation_extend_measurement ?

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+|===
+| Function Name                    | SBI Version | FID | EID
+| sbi_attestation_get_evidence     | TBD         | 0   | 0x41545354
+| sbi_measurement_extend           | TBD         | 1   | 0x41545354
```
/s/sbi_measurement_extend/sbi_attestation_extend_measurement ?

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+The reported capabilities allow attestation and measurement implementations to
+build platform agnostic frameworks.
+
+The SBI implementation places an attestation and measurement capabilities
```
places/writes ?

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+by a hash function that implements the algorithm reported by the
+`HASH_ALGORITHM` field from the <<get-capabilities-function>> returned structure
+(`ATST_CAPS`). See <<table_attestation_and_measurement_capabilities_format>> for
+more details.
```
As this function depends on get_capabilities function, that should be described first.

sameo replied:
I moved the `get_capabilities` function as the first FID, because the 2 other ones depend on it.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+separately. It can choose to maintain them through multiple measurement
+registers, or extend a single one for every `sbi_attestation_extend_measurement`
+call. In the former case, the caller can select a measurement index through the
+`measurement_index` parameter to extend one particular measurement register.
```
A few stupid question perhaps. 
How does the caller know which index to extend ?
Is measurement_index an optional parameter for the multiple measurement registers use case?
Are there any advantage for SBI implementation to choose one vs another ?

sameo replied:
> How does the caller know which index to extend ? 
> Is measurement_index an optional parameter for the multiple measurement registers use case? 

Good questions. I added a section that hopefully answers them.

> Are there any advantage for SBI implementation to choose one vs another ?

Not really. It's about how and if you want to separate dynamic/runtime measurements into different layers or not.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+formatted Certificate Signing Request (CSR) at the `cert_request_addr` physical
+address, which length is `cert_request_size`.
+
+Additionally, the caller provides a 64 bytes data blob at the
```
Along lines of Atish's normative text comment above, the "provides" can probably be a "must provide", especially since we refer to "must include a hash of the blob" below.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+The SBI implementation places an attestation and measurement capabilities
+structure at the caller's `caps_addr` physical address. The capabilities
+structure format follows the ÀTST_CAPS`format describe in
```
s/describe/described

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,334 @@ The possible error codes returned in `sbiret.error` are shown in the
+structure format follows the ÀTST_CAPS`format describe in
+<<table_attestation_and_measurement_capabilities_format>>.
+
+It is the caller's responsibility to allocate enough memory at `caps_addr` to
```
The caller must allocate...

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,337 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+[source, C]
+----
+struct sbiret sbi_attestation_get_capabilities(unsigned long caps_addr);
```
Should this be uint8_t* caps_addr?

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,337 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+[source, C]
+----
+struct sbiret sbi_attestation_get_evidence(unsigned long cert_request_addr,
```
Should pointer types should be uint8_t *? Also, pointers that are constant should have a const prefix.

atulkharerivos replied:
This needs to be marked as resolved.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1676,337 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+[source, C]
+----
+struct sbiret sbi_attestation_extend_measurement(unsigned long measurement_addr,
```
Same comment on pointer types.

sameo commented:
> Think I already mentioned this, but pointer types should be uint8_t * (and const uint8_t * where applicable). 

I agree we may want to express the fact that those are pointers, so I  changed them to `(const) unsigned long *`.
But looking at e.g. the new SBI debug console extension proposal, they are using `unsigned long` for specifiying physical address. I'm not sure what the convention is @atishp04?
Either way, the underlying implementation can do the casting according to what the argument is supposed to be, as long as it's clearly described.

> Also, can we pass explicitly pass in the size so that we don't have to worry about assumptions (like 136-bytes). Note that this is better for extensibility as well.

Yes, it's probably better. Done now, thanks.

sameo commented:
@rsahita @atishp04 @atulkharerivos I added one more function to this extension: `read_measurement`in order to read a measurement register back. That's mostly going to be needed by the TSM to retrieve all measurements from the TSM driver/RoT, in order to build the evidence certificate.

atishp04 commented:
> But looking at e.g. the new SBI debug console extension proposal, they are using `unsigned long` for specifiying physical address. I'm not sure what the convention is @atishp04?

It should be "unsigned long". That's the convention for all the extensions.

sameo commented:
> > But looking at e.g. the new SBI debug console extension proposal, they are using `unsigned long` for specifiying physical address. I'm not sure what the convention is @atishp04?
> 
> It should be "unsigned long". That's the convention for all the extensions.

Ok, I changed all the `unsigned long *` instances back to `unsigned long`. Thanks for the review.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+.Attestation and Measurement Capabilities Format Definition (ATST_CAPS)
+[cols="5,2,2,2,5", width=90%, align="center", options="header"]
+|===
+| Field Name            | Offset         | Type           | Size                   | Description
```
It would be good to explicitly call out the size in bytes (similar to table_measurement_register_descriptor)

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+| HASH_ALGORITHM        | 4              | Bitmap         | 1                     a| Measurement hash algorithm:
+
+                                                                                     * 0: SHA-384
+                                                                                     * 1: SHA-512
```
All other values are reserved for now?

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+
+                                                                                     * Bit 0: https://trustedcomputinggroup.org/wp-content/uploads/TCG_DICE_Attestation_Architecture_r22_02dec2020.pdf[`DiceTcbInfo`]
+                                                                                     * Bit 1: https://trustedcomputinggroup.org/wp-content/uploads/TCG_DICE_Attestation_Architecture_r22_02dec2020.pdf[`DiceMultiTcbInfo`]
+                                                                                     * Bit 2: https://pigweed.googlesource.com/open-dice/[`OpenDice`]
```
As this is a bitmap, you should point out that all other bits are reserved for now.

atulkharerivos replied:
This needs to be marked as resolved. It does appear that the bitmap was added below.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+----
+struct sbiret sbi_attestation_get_evidence(unsigned long cert_request_addr,
+                                           unsigned long cert_request_size
+                                           unsigned long request_data_addr,
```
nitpick: /s/request_data_addr/data_request_addr - Just to have similar format for argument names

atulkharerivos replied:
This needs to be marked as resolved.

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+| SBI_INVALID_PARAM   | The Certificate Signinig Request at `cert_request_addr`
+                        is invalid or `evidence_format` is not one of the
+                        evidence formats supported by the platform.
+| SBI_ERR_BUSY        | The evidence could not be generated due to some
```
What is the use case for this ? I am just trying to understand if SBI_ERR_BUSY can be applicable for other functions ?
It looks like the caller can immediately try if the the error code is SBI_ERR_BUSY. Is that the intention ?

atulkharerivos replied:
This needs to be marked as resolved (assume the retried text is new).

atishp04 commented on riscv-sbi.adoc:
```diff
@@ -1674,6 +1677,388 @@ The possible error codes returned in `sbiret.error` are shown in the
+----
+struct sbiret sbi_attestation_get_capabilities(unsigned long *caps_addr
+                                               unsigned long caps_size);
+----
```
Is it changed already ? Not sure if I am looking at the old PR.
It would also be good to prefix all the argument that (e.g caps_addr) updated by the SBI implementation with a prefix or suffix (e.g out) to indicate that.

sameo replied:
> Is it changed already ? Not sure if I am looking at the old PR.

You were looking at the right PR, I forgot to fix that one. Sorry.

> It would also be good to prefix all the argument that (e.g caps_addr) updated by the SBI implementation with a prefix or suffix (e.g out) to indicate that.

Good suggestion. I added a `_out` suffix for all pointer parameters that the SBI implementation updates.

atulkharerivos replied:
This needs to be marked as resolved.

sameo commented:
I just added `EBUSY` as a potential error for the measurement read and extension functions as well.
@atishp04 I think I addressed all your comments now. Let me know if you have some more.

atishp04 commented:
Looks good to me.

atulkharerivos commented on riscv-sbi.adoc:
```diff
@@ -201,6 +202,8 @@ error codes.
 |  SBI_ERR_ALREADY_STARTED   | -7
 |  SBI_ERR_ALREADY_STOPPED   | -8
+|  SBI_ERR_BUSY              | -9
+|  SBI_ERR_OVERFLOW          | -10
```
From the PR review: A related question about the semantics (and the associated name). Since the response would overflow the passed in buffer, the presumably the passed-in input buffer was too small, so perhaps it should be InsufficientCapacity / BufferTooSmall, or something like that?

atulkharerivos reviewed (commented):
LGTM.

vlsunil committed:
Sort the "Terms and Abbreviations" section

The "Terms and Abbreviations" section should be alphabetically
sorted.

Reviewed-by: Andrew Jones <ajones@ventanamicro.com>
Reviewed-by: Atish Patra <atishp@rivosinc.com>
Signed-off-by: Sunil V L <sunilvl@ventanamicro.com>

vlsunil committed:
Introduce CPPC extension

ACPI defines the Collaborative Processor Performance Control (CPPC),
a mechanism which is an abstract and flexible mechanism for the
Operating System to collaborate with an entity in the platform to
manage the performance of the processors.

The SBI CPPC extension provides an abstraction to access CPPC
registers through SBI calls.

Reviewed-by: Andrew Jones <ajones@ventanamicro.com>
Reviewed-by: Atish Patra <atishp@rivosinc.com>
Signed-off-by: Sunil V L <sunilvl@ventanamicro.com>

atishp04 committed:
checkin the pdf verison of the spec.

Signed-off-by: Atish Patra <atishp@rivosinc.com>

atishp04 committed:
Update the PRS TG mailing list.

The SBI related discussion happens in PRS TG instead of
platform mailing list. Update the readme to reflect that.

Signed-off-by: Atish Patra <atishp@rivosinc.com>

sameo committed:
SBI: Attestation Extension

This patch add an attestation and measurement SBI extension (`ATST`).

The `ATST` extension defines a set of functions for SBI implementations
to provide attestation primitives, mostly related to the confidential
computing remote attestation flows.

Besides allowing SBI callers to request for attestation certificates,
this extension also give them the ability to dynamically extend the TCB
measurements set and read measurement registers back.

SBI implementations must expose their attestation and measurement
capabilities through the `ATST` FID #0: `get_capabilities`.

Signed-off-by: Samuel Ortiz <sameo@rivosinc.com>
